vars/runATH.groovy

```diff
@@ -0,0 +1,160 @@
+        if (!isLocalATH) {
+            echo 'Checking connectivity to ATH sources…'
+            sh "git ls-remote --exit-code -h ${athUrl}"
+        }
+        if (isVersionNumber) {
+            jenkinsURl = mirror + "war/${jenkins}/jenkins.war"
+        } else if (jenkins == "latest") {
+            jenkinsURl = mirror + "war/latest/jenkins.war"
+        } else if (jenkins == "latest-rc") {
+            jenkinsURl = mirror + "/war-rc/latest/jenkins.war"
+        } else if (jenkins == "lts") {
+            jenkinsURl = mirror + "war-stable/latest/jenkins.war"
+        } else if (jenkins == "lts-rc") {
+            jenkinsURl = mirror + "war-stable-rc/latest/jenkins.war"
+        }
+
+        echo 'Checking whether Jenkins WAR is available…'
+        sh "curl -ILf ${jenkinsURl}"
+        // Validation ended
+
+        // ATH
+        if (isLocalATH) { // Deal with already existing ATH sources
+            athSourcesFolder = athUrl - "file://"
+        } else {
+            dir(athSourcesFolder) {
+                checkout changelog: true, poll: false, scm: [$class           : 'GitSCM', branches: [[name: athRevision]],
+                                                             userRemoteConfigs: [[url: athUrl]]]
+            }
+        }
+        dir(athSourcesFolder) {
+            def uid = sh(script: "id -u", returnStdout: true)
+            def gid = sh(script: "id -g", returnStdout: true)
+            athContainerImage = docker.build('jenkins/ath', "--build-arg=uid='$uid' --build-arg=gid='$gid' -f src/main/resources/ath-container/Dockerfile .")
+            // We may need to run things in parallel later, so avoid several checkouts
+            stash name: "athSources"
+        }
+
+        // Jenkins war
+        sh("curl -o jenkins.war -L ${jenkinsURl}")
+        stash includes: '*.war', name: 'jenkinsWar'
+
+    }
+    // Workaround for https://issues.jenkins-ci.org/browse/JENKINS-27092
+    if (shouldStop) {
+        return
+    }
+    // End of workaround
+
+    stage("Running ATH") {
+        def testsToRun = metadata.tests?.join(",")
+        def categoriesToRun = metadata.categories?.join(",")
+        def browsers = metadata.browsers ?: ["firefox"]
+        def failFast = metadata.failFast ?: false
+        def rerunCount = metadata.rerunFailingTestsCount ?: 0
+        def localSnapshots = metadata.useLocalSnapshots ?: true
+
+        if (testsToRun == null && categoriesToRun == null) {
+            categoriesToRun = defaultCategory
+        }
+
+        def testingbranches = ["failFast": failFast]
+        for (browser in browsers) {
+            // TODO remove this if else block once other browser are supported by infra
+            if (browser == "firefox") {
+
+                def currentBrowser = browser
+                def containerArgs = "-v /var/run/docker.sock:/var/run/docker.sock -v \$HOME/.m2/repository:/home/ath-user/.m2/repository -e LOCAL_SNAPSHOTS=${localSnapshots} -e SHARED_DOCKER_SERVICE=true -u ath-user"
```

Comment:
The flag `-v \$HOME/.m2/repository:/home/ath-user/.m2/repository` will cause a lot of trouble across developers and CI setups, because it not portable at all (cf. https://docs.docker.com/storage/#choose-the-right-type-of-mount).
=> It expects that the path `$HOME/.m2/repository` is:

* Consistent across setups
* Always available on the file system of the machine where the Docker Engine is run. Which won't be the case with the "remote docker engine pattern (when you bind-mount `/var/run/docker.sock` or when you spawn a transient VM for hosting such an engine).

=> Those 2 assumptions are false, making this non portable.

Proposal, as we discussed this together @raul-arabaolaza , it is use an handy feature of docker, the "volumes" (cf. https://docs.docker.com/storage/volumes/).

Instead of providing an absolute file system path as source (on the left of the ':' on the string supplied to the `-v` flag), we will provide a "docker volume name":
`-v ath-maven-local-repo:/home/ath-user/.m2/repository`

* If the named volume that not exists when `docker run..` is invoked, then it is created on-the-fly by the Docker Engine, which will manage and store this data on the machine where it is running (which might be different than the machine where `docker` command is invoked).
* If the named volumes already exists, then it will be reused (which is your goal here as I understand it).



There are 2 attention points on my proposal:
* The name of the volume is a convention. But it MUST be documented for the end users whom will call this groovy code. It SHOULD be set as an argument with a default value to the `call()` function, and then the resulting variable is supplied to the -v flag. Then you provide an handy extension point to the actor in charge of populating a maven repo prior to call the ATH.
* Permissions on the volume: you cannot be 100% sure about the permission inside this maven local repo. Since Docker can "root", I propose to add a container call that will ensure the permission set is on the expected state. This would be the following code to run BEFORE the first `athContainerImage.inside()` invokation (~line 127):
```
echo "${dockerVolumeNameForMavenLocalRepo}" // We expect this variable to be set with the volume name
sh "docker run -v ${dockerVolumeNameForMavenLocalRepo}:/repo -u root --entrypoint chown ${athContainerImage} -R /repo ath-user" // the "ath-user is already known inside your built image, it is the convention so let's assume this is the default
```

Reply:
Thanks for this @dduportal !!

Reply:
The named volume looks like the same approach followed in the [PCT docker image](https://github.com/jenkinsci/plugin-compat-tester) and it feels quite interesting, however, I need somehow to make sure that the node which actually builds the plugin ("linux" and "windows" in this case) use the aforementioned volume as storage and we may have version collisions if several jobs are running in parallel using the same maven repo...

Reply:
Right, but this is not related to this groovy file area of responsibilities. This is why I proposed to pass the volume name as an argument to the call method.
=> Then your Pipeline calling this groovy will play its role of "glue" by:

1. Generate a name for the docker volume, with enough discrimination to allow parallel builds and avoids conflicts on a single Docker Engine, and eventually create it (and ensure it is empty on the beginning of the build)
2. Pass this parameters to your "build plugin" function, to be sure the volume will be populated with the expected Maven cache. 
3. Pass this parameter when to the "runAth" step to have it reuse the volume.

N.B. for training, we use `"${GIT_BRANCH}-${GIT_COMMIT}"` for discrimination: it allows concurrents builds on a single Docker Engine, with still some "human readable" for when it is time to purge (example of policy: "if is does not start by "master-" and it is older than 5 days, then delete it).